library/core/src/ptr/mod.rs

```diff
@@ -1884,6 +1884,20 @@ pub unsafe fn write_volatile<T>(dst: *mut T, src: T) {
 ///
 /// Any questions go to @nagisa.
 #[lang = "align_offset"]
+#[safety::requires(a.is_power_of_two())]
+// If T is a ZST, then answer should either be 0 or usize::MAX
+#[safety::ensures(|result|
+    mem::size_of::<T>() != 0 || if p.addr() % a == 0 { *result == 0 } else { *result == usize::MAX}
+)]
+#[safety::ensures(|result| {
+    // If T is not a ZST and p can be aligned, then applying result as an offest should produce an aligned address
+    mem::size_of::<T>() == 0 || *result == usize::MAX ||
```

Comment:
Yes, it's deliberate--the contract is meant to say that if `align_offset` returns an answer (i.e., not `usize::MAX`) then the answer is correct.
I added a contract for one of the two non-ZST `usize::MAX` cases. I can't write a correctness contract for the last case (by last, I mean the one in the last line of `align_offset`, since it involves computing the gcd.

Reply:
The question is: if someone were to change the implementation of the function to just always return `usize::MAX`, do we want the proof harness to succeed?

Reply:
I changed the postcondition so that's no longer the case (and tested that verification does indeed fail when the function returns `usize::MAX`).